x/superfluid/keeper/stake_test.go

```diff
@@ -1028,6 +1031,522 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+		"lock that is superfluid delegated": {},
+		"lock that is superfluid undelegating": {
+			unlocking:              true,
+			superfluidUndelegating: true,
+		},
+		"lock that is unlocking": {
+			unlocking:              true,
+			superfluidUndelegating: false,
+		},
+		// error cases
+		"error: min amount to stake greater than actual amount": {
+			useMinAmountToStake: true,
+			expectedError:       true,
+		},
+		"error: use partial shares": {
+			usepartialShares: true,
+			expectedError:    true,
+		},
+		"error: use different lock ": {
+			usepartialShares: true,
+			expectedError:    true,
+		},
+		"error: sender is not owner of lock ": {
+			senderIsNotOwnerOfLock: true,
+			expectedError:          true,
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+			s.SetupTest()
+			s.Ctx = s.Ctx.WithBlockTime(defaultJoinTime)
+			// We bundle all migration setup into a single function to avoid repeating the same code for each test case.
+			_, _, lock, _, _, _, _, originalValAddr := s.SetupUnbondConvertAndStakeTest(s.Ctx, true, tc.superfluidUndelegating, false, false)
+
+			// testing params
+			sender := sdk.MustAccAddressFromBech32(lock.Owner)
+			if tc.senderIsNotOwnerOfLock {
+				sender = s.TestAccs[1]
+			}
+			valAddr := s.SetupValidator(stakingtypes.Bonded)
+			sharesToStake := lock.Coins
+			if tc.usepartialShares {
+				sharesToStake[0].Amount = lock.Coins[0].Amount.Quo(sdk.NewInt(2))
+			}
+			minAmountToStake := sdk.ZeroInt()
+			if tc.useMinAmountToStake {
+				minAmountToStake = sdk.NewInt(999999999)
+			}
+
+			balanceBeforeConvertLockToStake := s.App.BankKeeper.GetAllBalances(s.Ctx, sender)
+
+			// system under test
+			totalAmtConverted, sharesDelegated, err := s.App.SuperfluidKeeper.ConvertLockToStake(s.Ctx, sender, valAddr.String(), lock.ID, sharesToStake[0], minAmountToStake)
+			if tc.expectedError {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+
+			// Staking & Delegation check
+			// check if old delegation is succesfully deleted
+			_, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, originalValAddr)
+			s.Require().False(found)
+			// check if delegation amount matches
+			delegation, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, valAddr)
+			s.Require().True(found)
+			s.Require().Equal(delegation.Shares, sharesDelegated)
+			s.Require().True(totalAmtConverted.ToDec().Equal(delegation.Shares))
+
+			// Superfluid check
+			// The synthetic lockup should be deleted.
+			_, err = s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.StakingSyntheticDenom(lock.Coins[0].Denom, valAddr.String()))
+			s.Require().Error(err)
+
+			// Lock check
+			_, err = s.App.LockupKeeper.GetLockByID(s.Ctx, lock.ID)
+			s.Require().Error(err)
+
+			// Bank check
+			balanceAfterConvertLockToStake := s.App.BankKeeper.GetAllBalances(s.Ctx, sender)
+			s.Require().True(balanceBeforeConvertLockToStake.IsEqual(balanceAfterConvertLockToStake))
+		})
+	}
+}
```

Comment:
I know there's alot of duplication of these validation checks after testing these methods, but this was intentional due to 
- test cases differs a little for each cases
- code duplication  < visibility for test cases 
but open to other opinions